### ISSUE TITLE
local plugins

### ISSUE THREAD

**Issue by boneskull** [eslint/eslint#1651]
This ticket references an "open issue" mentioned in #1106.

Since #1386 was closed, if you:

- Want to write a plugin or rule, and
- Don't want to publish it to NPM, and
- Are using `eslint` in a non-CLI context, then
- Your tool/IDE must explicitly support the `--rulesdir` CLI option, or
- You are SOL.

Which happens to be the situation I'm in right now.  :smile: 

I'm using ESLint from a JetBrains IDE (which provides ESLint errors/warnings in context), and it simply reads a specific config file--it has no options for flags.  I imagine this is the case in many other tools as well.

I want to write a custom plugin, but don't want to bother with publishing to NPM.  I wanted to bang something out, but since I can't specify a `rulesdir` from my config file, I don't think I have any recourse.

**Comment by lo1tuma** [comment]
You can install your local plugin with npm and without publishing it to the public registry:

```sh
npm install /local-path/to/your/custom-plugin
```

`npm link` would work too.

**Comment by boneskull** [comment]
@lo1tuma So I'd have to do something like have project dir "foo" and "foo/my-plugin" then "npm --save-dev my-plugin" from "foo"??  Does npm allow that?  Seems pretty weird, if not bad/evil.

@nzakas I don't see what the resistance is to making this a bit easier for users, unless it's difficult to implement or would be a maintenance burden. Plugins are cool and all, but unless you can give your config a path to a plugin (not just a module name) it makes them a bit awkward for specific requirements.

**Comment by lo1tuma** [comment]
@boneskull You have to run `npm install ./my-plugin` from `foo/`. Local paths within your package.json [works since npm v2.0.0](https://docs.npmjs.com/files/package.json#local-paths).

**Comment by nzakas** [comment]
I think it's reasonable to limit the number of ways to do the same thing. `--rulesdir` was a quick hack to get runtime rules support until we got plugins.

There's already a way to load local plugins, so it appears your use case is already solved.

**Comment by boneskull** [comment]
Well, if you can `npm install` a subdirectory, then I'll do that.
